File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import deque
import sys

N,M,V=map(int,sys.stdin.readline().split())

def dfs(start):
    visited[start]=True
    result.append(start)
    
    for next in maps[start]:
        if not visited[next]:
            dfs(next)

def bfs(start):
    dq=deque()
    dq.append(start)
    visited[start]=True
    
    while dq:
        current=dq.popleft()
        result.append(current)
        
        for next in maps[current]:
            if not visited[next]:
                visited[next]=True
                dq.append(next)
        

    
maps=[[] for _ in range(N+1)]
for _ in range(M):
    a,b=map(int,sys.stdin.readline().split())
    maps[a].append(b)
    maps[b].append(a)
    
for i in range(N+1):
    maps[i].sort()
    
visited=[False]*(N+1) #방문
result=[]

dfs(V)
print(*result)

visited=[False]*(N+1)
result=[]
bfs(V)
print(*result)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.util.*;

public class Main{

    private static boolean[] visited;
    private static ArrayDeque<Integer> dq=new ArrayDeque<>(); //bfs
    private static ArrayList<Integer>[] adList; //dfs
    private static int N,M,V;
    private static StringBuilder sb=new StringBuilder();
    
    public static void main(String[] args){
        Scanner sc=new Scanner(System.in);
        N=sc.nextInt();
        M=sc.nextInt();
        V=sc.nextInt();
        
        visited=new boolean[N+1];
        adList=new ArrayList[N+1];
        for(int i=0;i<=N;i++){
            adList[i]=new ArrayList<>();
        }
        
        for(int i=0;i<M;i++){
            int x=sc.nextInt();
            int y=sc.nextInt();
            adList[x].add(y);
            adList[y].add(x);
        }
        
        for(int i=0;i<=N;i++){
            Collections.sort(adList[i]);
        }
        
        
        dfs(V);
        sb.append("\n");
        visited=new boolean[N+1];
        bfs(V);
        
        System.out.println(sb.toString());
    }
    
    private static void dfs(int start){
        visited[start]=true;
        sb.append(start+" ");
        
        for(int next:adList[start]){
            if(!visited[next]){
                dfs(next);
            }
        }
    }
    
    private static void bfs(int start){
        dq.offer(start);
        visited[start]=true;
        
        while(!dq.isEmpty()){
            int n=dq.poll();
            sb.append(n+" ");
            
            for(int next:adList[n]){
                if(!visited[next]){
                    dq.offer(next);
                    visited[next]=true;
                    
                }
            }
            
        }
    }
}